tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/types/registry/TensorTypeRegistry.java

```diff
@@ -0,0 +1,104 @@
+/*
+ *  Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *  =======================================================================
+ */
+package org.tensorflow.internal.types.registry;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.tensorflow.TensorMapper;
+import org.tensorflow.proto.framework.DataType;
+import org.tensorflow.types.TBfloat16;
+import org.tensorflow.types.TBool;
+import org.tensorflow.types.TFloat16;
+import org.tensorflow.types.TFloat32;
+import org.tensorflow.types.TFloat64;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.types.TInt64;
+import org.tensorflow.types.TString;
+import org.tensorflow.types.TUint8;
+import org.tensorflow.types.annotation.TensorType;
+import org.tensorflow.types.family.TType;
+
+/**
+ * Repository of all registered tensor types.
+ */
+public final class TensorTypeRegistry {
+
+  /**
+   * Find registered information about a tensor type from its equivalent data type
+   *
+   * @param dataType data type
+   * @return type registered information
+   * @throws IllegalArgumentException if no tensor type for this data type has been registered
+   */
+  public static <T extends TType> TensorTypeInfo<T> find(DataType dataType) {
```

Comment:
This is a place where it will infer the type bound from the calling context, and if that's wrong then we'll get a weird class cast error when people use it. It might be better to return the wildcard as at least the user will get a warning when they make the cast.

Reply:
Again note that the `TensorTypeRegistry` is (for now) an internal class so we do have some control over the context it is being called. But like you've suggested, let's review this later.

Reply:
Yeah, I know. But I'm in favour of having the compiler warn me before I make a silly mistake, and 6 months from now I won't necessarily remember that this method doesn't quite live up to its contract.